### PR TITLE
Extract MisnamedModuleInterfaceParserError from Flow and Typescript into error-utils.js

### PR DESCRIPTION
**Pull request by mohitcharkha**
## Summary

This PR is part of #34872 
This PR extracts MisnamedModuleFlowInterfaceParserError exception to a separate function inside an error-utils.js file

## Changelog

[Internal] [Changed] - Extract MisnamedModuleInterfaceParserError to a seperate function inside error-utils.js

## Test Plan

yarn jest react-native-codegen
Added unit case in `error-utils-test.js` file

Extract MisnamedModuleInterfaceParserError test Screenshot

### CODE REVIEW

**Comment by AntoineDoubovetzky** [comment]
@mohitcharkha I don't think the goal of this task was to extract the error itself, it is done for all errors here: https://github.com/facebook/react-native/pull/34896

The goal is to extract the logic that makes the check and throw if necessary (see https://github.com/facebook/react-native/issues/34872#issuecomment-1272930315)

File: packages/react-native-codegen/src/parsers/flow/modules/index.js [hunk 1]
```diff
@@ -639,9 +639,10 @@ function buildModuleSchema(
   const [moduleSpec] = moduleSpecs;
 
   if (moduleSpec.id.name !== 'Spec') {
-    throw new MisnamedModuleFlowInterfaceParserError(
+    throw new MisnamedModuleInterfaceParserError(
       hasteModuleName,
       moduleSpec.id,
+      'Flow',
     );
```
Comment (cipolleschi) [inline packages/react-native-codegen/src/parsers/flow/modules/index.js:646]: The tasks was slightly different: this PR export just the type from the `index.js` to the `error-utils`.
However, this step was responsibility of #34896 (which this PR should depend on).

What I was expecting here, was a function like:
```
function throwIfModuleInterfaceIsMisnamed(nativeModuleName: string, module: ASTNode, parserType: ParserType) {
  if (moduleSpec.id.name !== 'Spec') {
    throw new error
  }
}
```

That will replace these 6 lines in the `flow/modules/index.js` and in the `typescript/modules/index.js`.

In this way:
1. if we need to update the logic that throws the error, we do it in a single place
2. the parser code is more readable and self documenting.

Could you update this with this suggestion?

Thank you so much! 🙏

File: packages/react-native-codegen/src/parsers/typescript/modules/index.js [hunk 2]
```diff
@@ -679,9 +679,10 @@ function buildModuleSchema(
   const [moduleSpec] = moduleSpecs;
 
   if (moduleSpec.id.name !== 'Spec') {
-    throw new MisnamedModuleTypeScriptInterfaceParserError(
+    throw new MisnamedModuleInterfaceParserError(
       hasteModuleName,
       moduleSpec.id,
+      'TypeScript',
     );
   }
```
Comment (cipolleschi) [inline packages/react-native-codegen/src/parsers/typescript/modules/index.js:687]: Same change described above! 😉

File: packages/react-native-codegen/src/parsers/error-utils.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * @flow strict-local
+ * @format
+ */
+
+'use strict';
+
+const {ParserError} = require('./errors');
+
+type ParserTypes = 'Flow' | 'TypeScript';
```
Comment (cipolleschi) [inline packages/react-native-codegen/src/parsers/error-utils.js:15]: I love this! 😄  I'm moving the suggestion to the other PR (#34896).

**Comment by cipolleschi** [comment]
Also, given that we are introducing a new function, it would be great to have a unit test for it!

**Review by cipolleschi — APPROVED** [review]
Looks good! Thank you so much for taking this task!

File: packages/react-native-codegen/src/parsers/__tests__/error-utils-test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @format
+ * @oncall react_native
```
Comment (cipolleschi) [inline packages/react-native-codegen/src/parsers/__tests__/error-utils-test.js:8]: Could you please enable flow in the test file? You can do it by adding this line in the license:

```suggestion
 * @flow strict-local
 * @oncall react_native
```

**Review by cipolleschi — CHANGES_REQUESTED** [review]

**Comment by cipolleschi** [comment]
@mohitcharkha could you rebase and resolve the conflicts, please?

File: packages/react-native-codegen/src/parsers/typescript/modules/index.js [hunk 5]
```diff
@@ -73,6 +72,8 @@ const {
   throwIfModuleTypeIsUnsupported,
   throwIfUnusedModuleInterfaceParserError,
   throwIfModuleInterfaceNotFound,
+  throwIfMoreThanOneModuleRegistryCalls,
```
Comment (mohitcharkha) [inline packages/react-native-codegen/src/parsers/typescript/modules/index.js:75]:
```suggestion
```

File: packages/react-native-codegen/src/parsers/flow/modules/index.js [hunk 6]
```diff
@@ -70,6 +69,8 @@ const {
 
 const {
   throwIfModuleInterfaceNotFound,
+  throwIfModuleInterfaceIsMisnamed,
+  throwIfMoreThanOneModuleRegistryCalls,
```
Comment (mohitcharkha) [inline packages/react-native-codegen/src/parsers/flow/modules/index.js:73]:
```suggestion
```

**Review by cipolleschi — APPROVED** [review]

**Pull request closed without merging**